BootloaderCommonPkg/Library/Ext23Lib/Ext2Fs.c

```diff
@@ -577,58 +577,117 @@ SearchDirectory (
   )
 {
   PEI_EXT_PRIVATE_DATA *PrivateData;
   UINT8 *Buffer;
-  EXT2FS Ext2Fs;
+  EXT2FS *Ext2Fs;
   UINT32 BufSize;
   RETURN_STATUS Rc;
   UINT32 SbOffset;
 
   Rc = 0;
   Buffer = NULL;
 
-  PrivateData = (PEI_EXT_PRIVATE_DATA*) File->FileDevData;
+  if (FsHandle == NULL) {
+    Rc = RETURN_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  PrivateData = (PEI_EXT_PRIVATE_DATA *)FsHandle;
 
   Buffer = AllocatePool ((PrivateData->BlockSize > SBSIZE) ? PrivateData->BlockSize : SBSIZE);
   if (Buffer == NULL) {
     Rc = EFI_OUT_OF_RESOURCES;
     goto Exit;
   }
 
-  Rc = DEV_STRATEGY (File->DevPtr) (File->FileDevData, F_READ,
-                                    SBOFF / PrivateData->BlockSize, PrivateData->BlockSize, Buffer, &BufSize);
+  if (File == NULL) {
+    Rc = BDevStrategy (PrivateData, F_READ,
+                       SBOFF / PrivateData->BlockSize, PrivateData->BlockSize, Buffer, &BufSize);
+  } else {
+    Rc = DEV_STRATEGY (File->DevPtr) (PrivateData, F_READ,
+                                      SBOFF / PrivateData->BlockSize, PrivateData->BlockSize,
+                                      Buffer, &BufSize);
+  }
+
   if (Rc != 0) {
     goto Exit;
   }
+
   SbOffset = (SBOFF < PrivateData->BlockSize) ? SBOFF : 0;
-  E2FS_SBLOAD ((VOID *)(&Buffer[SbOffset]), &Ext2Fs);
-  if (Ext2Fs.Ext2FsMagic != E2FS_MAGIC) {
-    Rc = EFI_INVALID_PARAMETER;
+  Ext2Fs = (EXT2FS *)(&Buffer[SbOffset]);
+  if (Ext2Fs->Ext2FsMagic != E2FS_MAGIC) {
+    DEBUG ((DEBUG_ERROR, "Unsupported FS magic: 0x%x\n", Ext2Fs->Ext2FsMagic));
+    Rc = EFI_UNSUPPORTED;
     goto Exit;
   }
-  if (Ext2Fs.Ext2FsRev > E2FS_REV1 ||
-      (Ext2Fs.Ext2FsRev == E2FS_REV1 &&
-       (Ext2Fs.Ext2FsFirstInode != EXT2_FIRSTINO ||
-        (Ext2Fs.Ext2FsInodeSize != 128 && Ext2Fs.Ext2FsInodeSize != 256) ||
-        Ext2Fs.Ext2FsFeaturesIncompat & ~EXT2F_INCOMPAT_SUPP))) {
+
+  if (Ext2Fs->Ext2FsRev > E2FS_REV1 ||
+      (Ext2Fs->Ext2FsRev == E2FS_REV1 &&
+       (Ext2Fs->Ext2FsFirstInode != EXT2_FIRSTINO ||
+        (Ext2Fs->Ext2FsInodeSize != 128 && Ext2Fs->Ext2FsInodeSize != 256) ||
+        Ext2Fs->Ext2FsFeaturesIncompat & ~EXT2F_INCOMPAT_SUPP))) {
     Rc = EFI_UNSUPPORTED;
     goto Exit;
   }
+
+  if (RExt2Fs != NULL) {
+    E2FS_SBLOAD ((VOID *)Ext2Fs, RExt2Fs);
+  }
+
+Exit:
+  if (Buffer != NULL) {
+    FreePool (Buffer);
+  }
+
+  return Rc;
+}
+
+/**
+  Read Superblock of the file.
+
+  @param[in]      File          File for which super block needs to be read.
+  @param[in/out]  FileSystem    Fs on which super block is computed.
+
+  @retval 0 if superblock compute is success
+  @retval other if error.
+**/
+RETURN_STATUS
+EFIAPI
+ReadSBlock (
+  IN      OPEN_FILE     *File,
+  IN OUT  M_EXT2FS      *FileSystem
+  )
+{
+  PEI_EXT_PRIVATE_DATA *PrivateData;
+  EXT2FS Ext2Fs;
+  RETURN_STATUS Rc;
+
+  Rc = 0;
+
+  PrivateData = (PEI_EXT_PRIVATE_DATA*) File->FileDevData;
+
+  Rc = Ext2SbValidate ((EFI_HANDLE)PrivateData, File, &Ext2Fs);

```

Comment:
If use  &FileSystem->Ext2Fs instead,   then line #691 can be removed.

Reply:
let me check the impact of directly using &FileSystem->Ext2Fs in error condition.

Reply:
yes, it's no impact, so will update accordingly. thanks for the carefully review